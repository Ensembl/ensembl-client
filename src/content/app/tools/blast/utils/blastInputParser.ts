/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ParsedInputSequence } from 'src/content/app/tools/blast/types/parsedInputSequence';

/**
 * Allowed inputs:
 *
 * - bare sequence
 * - bare sequence with numbered lines
 * - bare sequences separated by empty lines
 * - FASTA-formatted sequence
 * - FASTA-formatted multiple sequences (separated by header lines)
 * - FASTA-formatted sequences separated by empty lines
 *
 * Also:
 * - if a sequence contains spaces, remove them
 * - make sure all sequence characters are upper-cased
 */

export const parseBlastInput = (input: string) => {
  input = input.trim();

  const sequences: ParsedInputSequence[] = [];
  let newSequence = createEmptySequence();

  const lines = input.split('\n');

  lines.forEach((line) => {
    line = prepareLine(line);

    if (
      isSequenceSeparator(line) &&
      (newSequence.value || newSequence.header)
    ) {
      addToSequences(newSequence, sequences);
      newSequence = createEmptySequence();
    }

    if (line.startsWith('>')) {
      // this is a FASTA header line indicating the beginning of a new sequence
      line = line.replace(/^\>\s*/, '');
      newSequence.header = line; // remove the leading ">" from the header line
    } else {
      newSequence.value += line;
    }
  });

  addToSequences(newSequence, sequences);

  return sequences;
};

const prepareLine = (line: string) => {
  line = line.trim();

  if (line.startsWith('>')) {
    // this is a FASTA header; return it as is
    return line;
  }

  // remove all digits and spaces from the line
  line = line.replace(/[\d\s]+/g, '');
  return line.toUpperCase();
};

const isSequenceSeparator = (line: string) => {
  return line.startsWith('>') || line === '';
};

const createEmptySequence = (): ParsedInputSequence => ({ value: '' });

const addToSequences = (
  sequence: ParsedInputSequence,
  sequences: ParsedInputSequence[]
) => {
  if (sequence.value || sequence.header) {
    sequences.push(sequence);
  }
};
