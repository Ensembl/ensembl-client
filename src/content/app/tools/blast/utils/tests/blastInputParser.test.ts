/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import fs from 'fs';
import path from 'path';

import { parseBlastInput } from '../blastInputParser';

const readFile = (pathToFile: string) => {
  pathToFile = path.resolve(__dirname, pathToFile);
  return fs.readFileSync(pathToFile, 'utf-8');
};

describe('parseBlastInput', () => {
  describe('parsing FASTA input', () => {
    it('correctly parses a single sequence', () => {
      const input = readFile('./input-sequence-examples/001.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });

    it('correctly parses a file with multiple sequences', () => {
      const input = readFile('./input-sequence-examples/002.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });

    it('ignores blank lines in the input', () => {
      const input = readFile('./input-sequence-examples/003.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });

    it('clears white spaces from the input', () => {
      const input = readFile('./input-sequence-examples/004.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });
  });

  describe('parsing bare sequences', () => {
    it('correctly parses a single sequence', () => {
      const input = readFile('./input-sequence-examples/005.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });

    it('cleans white space in the beginning of a sequence line', () => {
      const input = readFile('./input-sequence-examples/006.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });
    it('does not remove non-letter characters from numbered sequence lines', () => {
      const input = readFile('./input-sequence-examples/007.txt');
      expect(parseBlastInput(input)).toMatchSnapshot();
    });
  });
});
