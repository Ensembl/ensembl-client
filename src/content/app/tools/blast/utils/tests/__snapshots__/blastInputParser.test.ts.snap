// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseBlastInput > parsing FASTA input > clears white spaces from the input 1`] = `
[
  {
    "header": "sp|P05067|A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens OX=9606 GN=APP PE=1 SV=3",
    "value": "MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTKTCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVGEFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFRGVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEEEADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPCRAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARDPVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQAKNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITALQAVPPRPRHVFNMLKKYVRAEQKDRQHTLKHFEHVRMVDPKKAAQIRSQVMTHLRVIYERMNQSLSLLYNVPAVAEEIQDEVDELLQKEQNYSDDVLANMISEPRISYGNDALMPSLTETKTTVELLPVNGEFSLDDLQPWHSFGADSVPANTENEVEPVDARPAADRGLTTRPGSGLTNIKTEEISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVIATVIVITLVMLKKKQYTSIHHGVVEVDAAVTPEERHLSKMQQNGYENPTYKFFEQMQN",
  },
]
`;

exports[`parseBlastInput > parsing FASTA input > correctly parses a file with multiple sequences 1`] = `
[
  {
    "header": "gi|1348912|gb|G26680|G26680 human STS STS_D11729.^Agi|1396336|gb|G27617|G27617 human STS SHGC-32648.",
    "value": "CGGACCAGACGGACACAGGGAGAAGCTAGTTTCTTTCATGTGATTGANATNATGACTCTACTCCTAAAAGGGAAAAANCAATATCCTTGTTTACAGAAGAGAAACAAACAAGCCCCACTCAGCTCAGTCACAGGAGAGANCACAGAAAGTCTTAGGATCATGANCTCTGAAAAAAAGAGAAACCTTATCTTTNCTTTGTGGTTCCTTTAAACACACTCACACACACTTGGTCAGAGATGCTGTGCTTCTTGGAAGCAAGGNCTCAAAGGCAAGGTGCACGCAGAGGGACGTTTGAGTCTGGGATGAAGCATGTNCGTATTATTTATATGATGGAATTTCACGTTTTTATGTNAAGCNTGACAACACCAGGCAGGTATGAGAGGAAAGCAAGGCCCGTCCATNGCTGTCCGTACNCTTACGGNTTGCTTGTNGGAGNCATTTNGGTATTGTTTGTTGTAANANCCAAAANGGGCTTTGGNNTGGNAAAAGGGCAGANNGGGGGGGTTGGTGTNGTTTTTTGGGGGGANNNTTTNGATTTGGTNCCGGGNTTTNGTTTNCCNCGGNACCGGNTTTTGGTTGGGGNCCATTTNTGNGGGGCNTTGGNGTTNCNTTNCCCNNNTNNGANTGGTTTNA",
  },
  {
    "header": "gi|1348917|gb|G26685|G26685 human STS STS_D11734.",
    "value": "CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATGGCTGGGAATCTTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAGTTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAAGAGATTTACTGACTTTCCTAGAATAGTTTCTCTACTGGAAACCTGATGCTTTTATAAGCCATTGTGATTAGGATGACTGTTACAGGCTTAGCTTTGTGTGAAANCCAGTCACCTTTCTCCTAGGTAATGAGTAGTGCTGTTCATATTACTNTAAGTTCTATAGCATACTTGCNATCCTTTANCCATGCTTATCATANGTACCATTTGAGGAATTGNTTTGCCCTTTTGGGTTTNTTNTTGGTAAANNNTTCCCGGGTGGGGGNGGTNNNGAAA",
  },
  {
    "header": "gi|1592936|gb|G29385|G29385 human STS SHGC-32652",
    "value": "GATCAAATCTGCACTGTGTCTACATATAGGAAAGGTCCTGGTGTGTGCTAATGTTCCCAATGCAGGACTTGAGGAAGAGCTCTGTTATATGTTTCCATTTCTCTTTATCAAAGATAACCAAACCTTATGGCCCTTATAACAATGGAGGCACTGGCTGCCTCTTAATTTTCAATCATGGACCTAAAGAAGTACTCTGAAGGGTCTCAACAATGCCAGGTGGGGACAGATATACTCAGAGATTATCCAGGTCTGCCTCCCAGCGAGCCTGGAGTACACCAGACCCTCCTAGAGAAATCTGTTATAATTTACCACCCACTTATCCACCTTTAAACTTGGGGAAGGNNGCNTTTCAAATTAAATTTAATCNTNGGGGGNTTTTAAACTTTAACCCTTTTNCCNTTNTNGGGGTNGGNANTTGNCCCCNTTAAAGGGGGNNCCCCTNCNNGGGGGAATAAAACAANTTNNTTTTTT",
  },
]
`;

exports[`parseBlastInput > parsing FASTA input > correctly parses a single sequence 1`] = `
[
  {
    "header": "gi|3318709|pdb|1A91|  Subunit C Of The F1fo Atp Synthase Of Escherichia Coli; Nmr, 10 Structures",
    "value": "MENLNMDLLYMAAAVMMGLAAIGAAIGIGILGGKFLEGAARQPDLIPLLRTQFFIVMGLVDAIPMIAVGLGLYVMFAVA",
  },
]
`;

exports[`parseBlastInput > parsing FASTA input > ignores blank lines in the input 1`] = `
[
  {
    "header": "gi|1348912|gb|G26680|G26680 human STS STS_D11729.^Agi|1396336|gb|G27617|G27617 human STS SHGC-32648.",
    "value": "CGGACCAGACGGACACAGGGAGAAGCTAGTTTCTTTCATGTGATTGANATNATGACTCTACTCCTAAAAGGGAAAAANCAATATCCTTGTTTACAGAAGAGAAACAAACAAGCCCCACTCAGCTCAGTCACAGGAGAGANCACAGAAAGTCTTAGGATCATGANCTCTGAAAAAAAGAGAAACCTTATCTTTNCTTTGTGGTTCCTTTAAACACACTCACACACACTTGGTCAGAGATGCTGTGCTTCTTGGAAGCAAGGNCTCAAAGGCAAGGTGCACGCAGAGGGACGTTTGAGTCTGGGATGAAGCATGTNCGTATTATTTATATGATGGAATTTCACGTTTTTATGTNAAGCNTGACAACACCAGGCAGGTATGAGAGGAAAGCAAGGCCCGTCCATNGCTGTCCGTACNCTTACGGNTTGCTTGTNGGAGNCATTTNGGTATTGTTTGTTGTAANANCCAAAANGGGCTTTGGNNTGGNAAAAGGGCAGANNGGGGGGGTTGGTGTNGTTTTTTGGGGGGANNNTTTNGATTTGGTNCCGGGNTTTNGTTTNCCNCGGNACCGGNTTTTGGTTGGGGNCCATTTNTGNGGGGCNTTGGNGTTNCNTTNCCCNNNTNNGANTGGTTTNA",
  },
  {
    "header": "gi|1348917|gb|G26685|G26685 human STS STS_D11734.",
    "value": "CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATGGCTGGGAATCTTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAGTTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAAGAGATTTACTGACTTTCCTAGAATAGTTTCTCTACTGGAAACCTGATGCTTTTATAAGCCATTGTGATTAGGATGACTGTTACAGGCTTAGCTTTGTGTGAAANCCAGTCACCTTTCTCCTAGGTAATGAGTAGTGCTGTTCATATTACTNTAAGTTCTATAGCATACTTGCNATCCTTTANCCATGCTTATCATANGTACCATTTGAGGAATTGNTTTGCCCTTTTGGGTTTNTTNTTGGTAAANNNTTCCCGGGTGGGGGNGGTNNNGAAA",
  },
  {
    "header": "gi|1592936|gb|G29385|G29385 human STS SHGC-32652",
    "value": "GATCAAATCTGCACTGTGTCTACATATAGGAAAGGTCCTGGTGTGTGCTAATGTTCCCAATGCAGGACTTGAGGAAGAGCTCTGTTATATGTTTCCATTTCTCTTTATCAAAGATAACCAAACCTTATGGCCCTTATAACAATGGAGGCACTGGCTGCCTCTTAATTTTCAATCATGGACCTAAAGAAGTACTCTGAAGGGTCTCAACAATGCCAGGTGGGGACAGATATACTCAGAGATTATCCAGGTCTGCCTCCCAGCGAGCCTGGAGTACACCAGACCCTCCTAGAGAAATCTGTTATAATTTACCACCCACTTATCCACCTTTAAACTTGGGGAAGGNNGCNTTTCAAATTAAATTTAATCNTNGGGGGNTTTTAAACTTTAACCCTTTTNCCNTTNTNGGGGTNGGNANTTGNCCCCNTTAAAGGGGGNNCCCCTNCNNGGGGGAATAAAACAANTTNNTTTTTT",
  },
]
`;

exports[`parseBlastInput > parsing bare sequences > cleans white space in the beginning of a sequence line 1`] = `
[
  {
    "value": "QIKDLLVSSSTDLDTTLVLVNAIYFKGMWKTAFNAEDTREMPFHVTKQESKPVQMMCMNNSFNVATLPAEKMKILELPFASGDLSMLVLLPDEVSDLERIEKTINFEKLTEWTNPNTMEKRRVKVYLPQMKIEEKYNLTSVLMALGMTDLFIPSANLTGISSAESLKISQAVHGAFMELSEDGIEMAGSTGVIEDIKHSPESEQFRADHPFLFLIKHNPTNTIVYFGRYWSP",
  },
]
`;

exports[`parseBlastInput > parsing bare sequences > correctly parses a single sequence 1`] = `
[
  {
    "value": "QIKDLLVSSSTDLDTTLVLVNAIYFKGMWKTAFNAEDTREMPFHVTKQESKPVQMMCMNNSFNVATLPAEKMKILELPFASGDLSMLVLLPDEVSDLERIEKTINFEKLTEWTNPNTMEKRRVKVYLPQMKIEEKYNLTSVLMALGMTDLFIPSANLTGISSAESLKISQAVHGAFMELSEDGIEMAGSTGVIEDIKHSPESEQFRADHPFLFLIKHNPTNTIVYFGRYWSP",
  },
]
`;

exports[`parseBlastInput > parsing bare sequences > removes numbers from numbered sequence lines 1`] = `
[
  {
    "value": "QIKDLLVSSSTDLDTTLVLVNAIYFKGMWKTAFNAEDTREMPFHVTKQESKPVQMMCMNNSFNVATLPAEKMKILELPFASGDLSMLVLLPDEVSDLERIEKTINFEKLTEWTNPNTMEKRRVKVYLPQMKIEEKYNLTSVLMALGMTDLFIPSANLTGISSAESLKISQAVHGAFMELSEDGIEMAGSTGVIEDIKHSPESEQFRADHPFLFLIKHNPTNTIVYFGRYWSP",
  },
]
`;

exports[`parseBlastInput parsing FASTA input clears white spaces from the input 1`] = `
[
  {
    "header": "sp|P05067|A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens OX=9606 GN=APP PE=1 SV=3",
    "value": "MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTKTCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVGEFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFRGVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEEEADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPCRAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARDPVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQAKNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITALQAVPPRPRHVFNMLKKYVRAEQKDRQHTLKHFEHVRMVDPKKAAQIRSQVMTHLRVIYERMNQSLSLLYNVPAVAEEIQDEVDELLQKEQNYSDDVLANMISEPRISYGNDALMPSLTETKTTVELLPVNGEFSLDDLQPWHSFGADSVPANTENEVEPVDARPAADRGLTTRPGSGLTNIKTEEISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVIATVIVITLVMLKKKQYTSIHHGVVEVDAAVTPEERHLSKMQQNGYENPTYKFFEQMQN",
  },
]
`;

exports[`parseBlastInput parsing FASTA input correctly parses a file with multiple sequences 1`] = `
[
  {
    "header": "gi|1348912|gb|G26680|G26680 human STS STS_D11729.^Agi|1396336|gb|G27617|G27617 human STS SHGC-32648.",
    "value": "CGGACCAGACGGACACAGGGAGAAGCTAGTTTCTTTCATGTGATTGANATNATGACTCTACTCCTAAAAGGGAAAAANCAATATCCTTGTTTACAGAAGAGAAACAAACAAGCCCCACTCAGCTCAGTCACAGGAGAGANCACAGAAAGTCTTAGGATCATGANCTCTGAAAAAAAGAGAAACCTTATCTTTNCTTTGTGGTTCCTTTAAACACACTCACACACACTTGGTCAGAGATGCTGTGCTTCTTGGAAGCAAGGNCTCAAAGGCAAGGTGCACGCAGAGGGACGTTTGAGTCTGGGATGAAGCATGTNCGTATTATTTATATGATGGAATTTCACGTTTTTATGTNAAGCNTGACAACACCAGGCAGGTATGAGAGGAAAGCAAGGCCCGTCCATNGCTGTCCGTACNCTTACGGNTTGCTTGTNGGAGNCATTTNGGTATTGTTTGTTGTAANANCCAAAANGGGCTTTGGNNTGGNAAAAGGGCAGANNGGGGGGGTTGGTGTNGTTTTTTGGGGGGANNNTTTNGATTTGGTNCCGGGNTTTNGTTTNCCNCGGNACCGGNTTTTGGTTGGGGNCCATTTNTGNGGGGCNTTGGNGTTNCNTTNCCCNNNTNNGANTGGTTTNA",
  },
  {
    "header": "gi|1348917|gb|G26685|G26685 human STS STS_D11734.",
    "value": "CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATGGCTGGGAATCTTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAGTTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAAGAGATTTACTGACTTTCCTAGAATAGTTTCTCTACTGGAAACCTGATGCTTTTATAAGCCATTGTGATTAGGATGACTGTTACAGGCTTAGCTTTGTGTGAAANCCAGTCACCTTTCTCCTAGGTAATGAGTAGTGCTGTTCATATTACTNTAAGTTCTATAGCATACTTGCNATCCTTTANCCATGCTTATCATANGTACCATTTGAGGAATTGNTTTGCCCTTTTGGGTTTNTTNTTGGTAAANNNTTCCCGGGTGGGGGNGGTNNNGAAA",
  },
  {
    "header": "gi|1592936|gb|G29385|G29385 human STS SHGC-32652",
    "value": "GATCAAATCTGCACTGTGTCTACATATAGGAAAGGTCCTGGTGTGTGCTAATGTTCCCAATGCAGGACTTGAGGAAGAGCTCTGTTATATGTTTCCATTTCTCTTTATCAAAGATAACCAAACCTTATGGCCCTTATAACAATGGAGGCACTGGCTGCCTCTTAATTTTCAATCATGGACCTAAAGAAGTACTCTGAAGGGTCTCAACAATGCCAGGTGGGGACAGATATACTCAGAGATTATCCAGGTCTGCCTCCCAGCGAGCCTGGAGTACACCAGACCCTCCTAGAGAAATCTGTTATAATTTACCACCCACTTATCCACCTTTAAACTTGGGGAAGGNNGCNTTTCAAATTAAATTTAATCNTNGGGGGNTTTTAAACTTTAACCCTTTTNCCNTTNTNGGGGTNGGNANTTGNCCCCNTTAAAGGGGGNNCCCCTNCNNGGGGGAATAAAACAANTTNNTTTTTT",
  },
]
`;

exports[`parseBlastInput parsing FASTA input correctly parses a single sequence 1`] = `
[
  {
    "header": "gi|3318709|pdb|1A91|  Subunit C Of The F1fo Atp Synthase Of Escherichia Coli; Nmr, 10 Structures",
    "value": "MENLNMDLLYMAAAVMMGLAAIGAAIGIGILGGKFLEGAARQPDLIPLLRTQFFIVMGLVDAIPMIAVGLGLYVMFAVA",
  },
]
`;

exports[`parseBlastInput parsing FASTA input ignores blank lines in the input 1`] = `
[
  {
    "header": "gi|1348912|gb|G26680|G26680 human STS STS_D11729.^Agi|1396336|gb|G27617|G27617 human STS SHGC-32648.",
    "value": "CGGACCAGACGGACACAGGGAGAAGCTAGTTTCTTTCATGTGATTGANATNATGACTCTACTCCTAAAAGGGAAAAANCAATATCCTTGTTTACAGAAGAGAAACAAACAAGCCCCACTCAGCTCAGTCACAGGAGAGANCACAGAAAGTCTTAGGATCATGANCTCTGAAAAAAAGAGAAACCTTATCTTTNCTTTGTGGTTCCTTTAAACACACTCACACACACTTGGTCAGAGATGCTGTGCTTCTTGGAAGCAAGGNCTCAAAGGCAAGGTGCACGCAGAGGGACGTTTGAGTCTGGGATGAAGCATGTNCGTATTATTTATATGATGGAATTTCACGTTTTTATGTNAAGCNTGACAACACCAGGCAGGTATGAGAGGAAAGCAAGGCCCGTCCATNGCTGTCCGTACNCTTACGGNTTGCTTGTNGGAGNCATTTNGGTATTGTTTGTTGTAANANCCAAAANGGGCTTTGGNNTGGNAAAAGGGCAGANNGGGGGGGTTGGTGTNGTTTTTTGGGGGGANNNTTTNGATTTGGTNCCGGGNTTTNGTTTNCCNCGGNACCGGNTTTTGGTTGGGGNCCATTTNTGNGGGGCNTTGGNGTTNCNTTNCCCNNNTNNGANTGGTTTNA",
  },
  {
    "header": "gi|1348917|gb|G26685|G26685 human STS STS_D11734.",
    "value": "CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATGGCTGGGAATCTTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAGTTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAAGAGATTTACTGACTTTCCTAGAATAGTTTCTCTACTGGAAACCTGATGCTTTTATAAGCCATTGTGATTAGGATGACTGTTACAGGCTTAGCTTTGTGTGAAANCCAGTCACCTTTCTCCTAGGTAATGAGTAGTGCTGTTCATATTACTNTAAGTTCTATAGCATACTTGCNATCCTTTANCCATGCTTATCATANGTACCATTTGAGGAATTGNTTTGCCCTTTTGGGTTTNTTNTTGGTAAANNNTTCCCGGGTGGGGGNGGTNNNGAAA",
  },
  {
    "header": "gi|1592936|gb|G29385|G29385 human STS SHGC-32652",
    "value": "GATCAAATCTGCACTGTGTCTACATATAGGAAAGGTCCTGGTGTGTGCTAATGTTCCCAATGCAGGACTTGAGGAAGAGCTCTGTTATATGTTTCCATTTCTCTTTATCAAAGATAACCAAACCTTATGGCCCTTATAACAATGGAGGCACTGGCTGCCTCTTAATTTTCAATCATGGACCTAAAGAAGTACTCTGAAGGGTCTCAACAATGCCAGGTGGGGACAGATATACTCAGAGATTATCCAGGTCTGCCTCCCAGCGAGCCTGGAGTACACCAGACCCTCCTAGAGAAATCTGTTATAATTTACCACCCACTTATCCACCTTTAAACTTGGGGAAGGNNGCNTTTCAAATTAAATTTAATCNTNGGGGGNTTTTAAACTTTAACCCTTTTNCCNTTNTNGGGGTNGGNANTTGNCCCCNTTAAAGGGGGNNCCCCTNCNNGGGGGAATAAAACAANTTNNTTTTTT",
  },
]
`;

exports[`parseBlastInput parsing bare sequences cleans white space in the beginning of a sequence line 1`] = `
[
  {
    "value": "QIKDLLVSSSTDLDTTLVLVNAIYFKGMWKTAFNAEDTREMPFHVTKQESKPVQMMCMNNSFNVATLPAEKMKILELPFASGDLSMLVLLPDEVSDLERIEKTINFEKLTEWTNPNTMEKRRVKVYLPQMKIEEKYNLTSVLMALGMTDLFIPSANLTGISSAESLKISQAVHGAFMELSEDGIEMAGSTGVIEDIKHSPESEQFRADHPFLFLIKHNPTNTIVYFGRYWSP",
  },
]
`;

exports[`parseBlastInput parsing bare sequences correctly parses a single sequence 1`] = `
[
  {
    "value": "QIKDLLVSSSTDLDTTLVLVNAIYFKGMWKTAFNAEDTREMPFHVTKQESKPVQMMCMNNSFNVATLPAEKMKILELPFASGDLSMLVLLPDEVSDLERIEKTINFEKLTEWTNPNTMEKRRVKVYLPQMKIEEKYNLTSVLMALGMTDLFIPSANLTGISSAESLKISQAVHGAFMELSEDGIEMAGSTGVIEDIKHSPESEQFRADHPFLFLIKHNPTNTIVYFGRYWSP",
  },
]
`;

exports[`parseBlastInput parsing bare sequences removes numbers from numbered sequence lines 1`] = `
[
  {
    "value": "QIKDLLVSSSTDLDTTLVLVNAIYFKGMWKTAFNAEDTREMPFHVTKQESKPVQMMCMNNSFNVATLPAEKMKILELPFASGDLSMLVLLPDEVSDLERIEKTINFEKLTEWTNPNTMEKRRVKVYLPQMKIEEKYNLTSVLMALGMTDLFIPSANLTGISSAESLKISQAVHGAFMELSEDGIEMAGSTGVIEDIKHSPESEQFRADHPFLFLIKHNPTNTIVYFGRYWSP",
  },
]
`;
